include/deal.II/integrators/grad_div.h

```diff
@@ -0,0 +1,379 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2010 - 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__integrators_grad_div_h
+#define dealii__integrators_grad_div_h
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/quadrature.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/fe/mapping.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/meshworker/dof_info.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LocalIntegrators
+{
+  /**
+   * Local integrators related to the grad-div operator and its boundary
+   * traces
+   *
+   * @ingroup Integrators
+   * @author Guido Kanschat, Timo Heister
+   * @date 2016
+   */
+  namespace GradDiv
+  {
+    /**
+     * The weak form of the grad-div operator penalizing volume changes
+     * @f[
+     *  \int_Z \nabla\!\cdot\!u \nabla\!\cdot\!v \,dx
+     * @f]
+     *
+     * @author Guido Kanschat
+     * @date 2011
+     */
+    template <int dim>
+    void cell_matrix (
+      FullMatrix<double> &M,
+      const FEValuesBase<dim> &fe,
+      double factor = 1.)
+    {
+      const unsigned int n_dofs = fe.dofs_per_cell;
+
+      AssertDimension(fe.get_fe().n_components(), dim);
+      AssertDimension(M.m(), n_dofs);
+      AssertDimension(M.n(), n_dofs);
+
+      for (unsigned int k=0; k<fe.n_quadrature_points; ++k)
+        {
+          const double dx = factor * fe.JxW(k);
+          for (unsigned int i=0; i<n_dofs; ++i)
+            for (unsigned int j=0; j<n_dofs; ++j)
+              {
+                double dv = 0.;
+                double du = 0.;
+                for (unsigned int d=0; d<dim; ++d)
+                  {
+                    dv += fe.shape_grad_component(i,k,d)[d];
+                    du += fe.shape_grad_component(j,k,d)[d];
+                  }
+
+                M(i,j) += dx * du * dv;
+              }
+        }
+    }
+
+    /**
+     * The weak form of the grad-div residual
+     * @f[
+     *  \int_Z \nabla\cdot u \nabla \cdot v \,dx
```

Comment:
They are all written such that the matrix is the derivative of the residual without sign change.

Residual refers to the Newton residual in nonlinear computations, and indeed it had never occurred to me that it is defined the other way round in linear algebra.